core/src/main/java/io/grpc/internal/DelayedStream.java

```diff
@@ -104,19 +119,40 @@ public void run() {
   /**
    * Transfers all pending and future requests and mutations to the given stream.
    *
-   * <p>No-op if either this method or {@link #cancel} have already been called.
+   * <p>This method must be called at most once.  Extraneous calls will throw and end up cancelling
+   * the given streams.
+   *
+   * <p>If {@link #cancel} has been called, this method will cancel the given stream.
    */
   // When this method returns, passThrough is guaranteed to be true
   final void setStream(ClientStream stream) {
+    ClientStream savedRealStream;
+    Status savedError;
     synchronized (this) {
-      // If realStream != null, then either setStream() or cancel() has been called.
-      if (realStream != null) {
-        return;
+      savedRealStream = realStream;
+      savedError = error;
+      if (savedRealStream == null) {
+        realStream = checkNotNull(stream, "stream");
       }
-      realStream = checkNotNull(stream, "stream");
     }
 
-    drainPendingCalls();
+    if (savedRealStream == null) {
+      drainPendingCalls();
+    } else {
+      // If realStream was not null, then either setStream() or cancel() must had been called,
+      // we will cancel and discard the given stream.
+      // ClientStream.cancel() must be called after start()
+      stream.start(NOOP_STREAM_LISTENER);
+      if (savedError != null) {
+        stream.cancel(savedError);
```

Comment:
Let's not cancel with `savedError`. We're cancelling the stream because of the incorrect API usage, not because of a previous error.

At that point `error` itself can be deleted.